BLUEBUTTON-1192: Verify database queries

jcgeer opened the pull request:
First draft of bash script to verify database queries

jcgeer committed:
BLUEBUTTON-1192: Verify database queries

First draft of bash script to verify database queries

karlmdavis commented on dev/verify-db-queries.sh:
```diff
@@ -0,0 +1,124 @@
+# Constants
+# FIXME figure out DB name/verify username
+dbname = "DATABASE_NAME"
+username = "postgres"
```
Please make `dbname`, `username` and other variables that are either security sensitive or variable, either arguments or environment variables, rather than a constant.

karlmdavis commented on dev/verify-db-queries.sh:
```diff
@@ -0,0 +1,124 @@
+	# run all DB queries once
+	
+	# database_query.bene_by_hicn.bene_by_hicn_or_id.include_hicns_and_mbis.query
+	`psql -U $username -d $dbname -c "select beneficiar0_.\"beneficiaryId\" as benefici1_0_0_, beneficiar1_.\"beneficiaryHistoryId\" as benefici1_1_1_, medicarebe2_.\"medicareBeneficiaryIdKey\" as medicare1_12_2_, beneficiar0_.\"ageOfBeneficiary\" as ageOfBen2_0_0_, beneficiar0_.\"beneEnrollmentReferenceYear\" as beneEnro3_0_0_, beneficiar0_.\"beneficiaryDateOfDeath\" as benefici4_0_0_, beneficiar0_ ...
```
Will this time out after 30 seconds? Or are you just expecting that Jo's script will kill it (which might take up to 120 seconds, FYI).

I'd probably use `statement_timeout`, per https://dba.stackexchange.com/a/164450.

jcgeer replied:
I've set it to 15 seconds that way earlier in the script, line 13. Do we want it at 30 seconds?

karlmdavis replied:
I missed that, but did you test it yet?

I'm pretty sure that `SET` statements like that are per-session, and each `psql` command you're running will be in a separate session.

jcgeer replied:
From reading further it looks like you're right, it is per-session. I didn't realize each `psql` command would be in a separate session. I've added it into a new function to perform before each statement is run.

karlmdavis commented on dev/verify-db-queries.sh:
```diff
@@ -0,0 +1,124 @@
+	fi
+	
+	# database_query.bene_by_hicn.bene_by_hicn_or_id.omit_hicns_and_mbis.query
+	`psql -U $username -d $dbname -c "select beneficiar0_.\"beneficiaryId\" as benefici1_0_, beneficiar0_.\"ageOfBeneficiary\" as ageOfBen2_0_, beneficiar0_.\"beneEnrollmentReferenceYear\" as beneEnro3_0_, beneficiar0_.\"beneficiaryDateOfDeath\" as benefici4_0_, beneficiar0_.\"birthDate\" as birthDat5_0_, beneficiar0_.\"countyCode\" as countyCo6_0_, beneficiar0_.\"currentBeneficiaryIdCode\" as curre ...
```
I'd probably write a Bash function for all of your calls to `psql`, so you can reuse logic, per https://www.tldp.org/LDP/abs/html/functions.html.

karlmdavis commented on dev/verify-db-queries.sh:
```diff
@@ -0,0 +1,124 @@
+	if [ echo $? != 0]
+	then
+		timedout = true
+	fi
```
You'll need to do this for every query, which is another good reason to build a reuseable function (see below).

karlmdavis commented on dev/verify-db-queries.sh:
```diff
@@ -0,0 +1,124 @@
+randomBeneIds = `psql -U $username -d $dbname -c "SELECT beneficiaryId FROM Beneficiaries ORDER BY RANDOM() LIMIT $numBenesToQuery"`
+
+# set queries to timeout after 15 seconds
+`psql -U $username -d $dbname -c "SET statement_timeout TO 15000"`
```
FYI: You don't actually need the backticks unless you're assigning the output to a variable.

karlmdavis reviewed (changes requested):
Great start! Some suggestions/fixes, but they're all rather minor.

jcgeer committed:
Some updates from the first review

-Breaking repeated code out into functions
-Some formatting (removing back ticks, putting db queries into separate
txt file, etc.)

jcgeer committed:
More review updates

-Changing dbname, username, and numBenesToQuery to argument variables to
be adjusted as needed
-Adding comments

jcgeer committed:
Separating functions out a bit

BLUEBUTTON-1192

jcgeer committed:
Using correct syntax to pass function parameters

BLUEBUTTON-1192

jcgeer committed:
Fixing syntax for assignment statements

jcgeer committed:
Backticks depracated, switching to $(command)

BLUEBUTTON-1192

jcgeer committed:
Missed an ending parenthesis

BLUEBUTTON-1192

jcgeer committed:
Changing randomBeneIds to an array

How randomBeneIds was originally assigned was not the right syntax for
creating an array or looped through correctly. Changed that, as well as
a couple other minor errors.

BLUEBUTTON-1192

karlmdavis commented on dev/verify-db-queries.sh:
```diff
@@ -0,0 +1,70 @@
+databaseQueries="/bluebutton-ansible-playbooks-data/dev/database-queries.txt"
+
+# select random sample of Beneficiaries.beneficiaryId
+randomBeneIds=( $(psql -U $USERNAME -d $DBNAME -c "SELECT beneficiaryId FROM Beneficiaries ORDER BY RANDOM() LIMIT $NUMBENESTOQUERY") )
```
Why the extra parentheses here? Won't hurt anything, I don't think, but it causes the command to be run in a subshell.

https://www.tldp.org/LDP/abs/html/subshells.html

jcgeer replied:
Must have added those accidentally.

karlmdavis commented on dev/verify-db-queries.sh:
```diff
@@ -0,0 +1,70 @@
+
+runPsqlQuery() {
+	dbQuery=$1
+	psql -U $USERNAME -d $DBNAME $dbQuery
```
This still won't work with the timeout: you're still running it in a separate process, which means it's a separate session. Try this, instead (per the `psql` man page: http://manpages.ubuntu.com/manpages/bionic/man1/psql.1.html):

```suggestion
	psql -U $USERNAME -d $DBNAME <<EOF
SET statement_timeout TO 15000;
${dbQuery}
EOF
```

This odd `<<EOF` construct is called a "heredoc" in Bash, and they're weird, but useful: http://tldp.org/LDP/abs/html/here-docs.html.

jcgeer replied:
Thanks for the documentation on this. I wasn't sure how to do it otherwise, obviously.

karlmdavis reviewed (changes requested):
Still needs one change.

jcgeer committed:
Review changes

-Removing some extra parentheses
-Changing the set timeout statement to be done in the same session as
the query using a here-document.

jcgeer committed:
Adding curly braces

Adding the curly braces so that the command is executed in the current
shell context.

BLUEBUTTON-1192

karlmdavis commented on dev/verify-db-queries.sh:
```diff
@@ -31,7 +26,10 @@ checkForTimeout() {
+	psql -U $USERNAME -d $DBNAME <<EOF
+	SET statement_timeout TO 15000;
+	${dbQuery}
+	EOF
```
I think you need to remove the leading whitespace on the last three lines of this block, as the heredoc won't trim it -- it'll run those commands with the spaces, as written. I know it looks weird like that, but that's how they work.

(The spaces are probably fine in this context: I don't think SQL queries care, but it's best to not get in the habit.)

jcgeer replied:
Fair enough. I see I hadn't read that far down into the documentation yet.

karlmdavis reviewed (changes requested):
One last thing.

jcgeer committed:
Removing whitespace

This whitespace is unneeded according to the documentation for
here-docs.

BLUEBUTTON-1192

karlmdavis reviewed (approved):
Approved, though as mentioned: please verify on TEST that this works as expected with a load test before deploying to DPR and PROD.

Great work!

karlmdavis committed:
Reworked DB query verification script.

Got it to a point where it's incomplete, but does work.

BLUEBUTTON-1192

karlmdavis commented:
Went to run this myself out in TEST this weekend and realized it needed some work. Hope you don't mind, but just pushed my changes here.

It's now working.

Notes:

* Still not 100% done: I left a lot of queries out.
* May want to rework it further to run some or all of the queries within the same DB connection, as we're wasting a lot of time connecting, disconnecting, etc.
    * This is the same reason that we use a DB connection pool in our app code.
    * On the other hand, if we're just going to run everything in the same connection, we should consider using `pgbench`, instead.
* I ran this in TEST today for 100K benes, with only the carrier claims query enabled. It didn't find any timeouts.
    * Maybe TEST just doesn't have enough data?
    * Maybe there's something else about PROD that causes these problems to appear?
    * Maybe the problem only shows up when lots of other queries are also being run?
    * ... not sure. Now have a run going in TEST with all the query-by-bene-ID queries enabled.

jcgeer committed:
Adding additional queries

BLUEBUTTON-1192

jcgeer committed:
Fixing some copy/paste errors

BLUEBUTTON-1192

jcgeer committed:
Correcting PDE query to select "eventId"

BLUEBUTTON-1192

karlmdavis commented:
Closing this PR, as the project has now been migrated into a new monorepo: https://github.com/CMSgov/beneficiary-fhir-data
